syncapi/consumers/roomserver.go

```diff
@@ -622,6 +622,23 @@ func (s *OutputRoomEventConsumer) writeFTS(ev *rstypes.HeaderedEvent, pduPositio
 		if err := s.fts.Index(e); err != nil {
 			return err
 		}
+		// If the event is an edited message we remove the original event from the index
+		// to avoid duplicates in the search results.
+		relatesTo := gjson.GetBytes(ev.Content(), "m\\.relates_to")
+		if relatesTo.Exists() {
+			relatedData := relatesTo.Value().(map[string]interface{})
+			if _, ok := relatedData["rel_type"]; ok && relatedData["rel_type"] == "m.replace" {
+				// We remove the original event from the index
+				if srcEventID, ok := relatedData["event_id"]; ok {
+					if err := s.fts.Delete(srcEventID.(string)); err != nil {
+						log.WithFields(log.Fields{
+							"event_id": ev.EventID(),
+							"src_id":   srcEventID,
+						}).WithError(err).Error("Failed to delete edited message from the fulltext index")
+					}
+				}
+			}
+		}
```

Comment:
What do you think about:
```go
		relType := gjson.GetBytes(ev.Content(), "m\\.relates_to.rel_type").Str
		if relType == "m.replace" {
			eventID := gjson.GetBytes(ev.Content(), "m\\.relates_to.event_id").Str
			if err := s.fts.Delete(eventID); err != nil {
				log.WithFields(log.Fields{
					"event_id": ev.EventID(),
					"src_id":   eventID,
				}).WithError(err).Error("Failed to delete edited message from the fulltext index")
			}
		}
```

imo this is easier to read than type assertions (and probably less error prone)

Reply:
Indeed, i don't like assertions too, but I thought to avoid double parsing so didn't find another way to do this.

Reply:
We can also move this part of the code up to the `case "m.room.message"`, so we don't parse the event for `m.room.name` etc (even though that shouldn't make a huge difference, except when someone changes the topic/name continuously)

Reply:
Sure, I'll fix it shortly

Reply:
I've looked into it with fresh eyes and replaced type asserions with Map() method of gjson.Result. 
I've also moved parsing into "m.room.message" section leaving deletion on it's place as we can delete the source message only after the replacing message was indexed successfully.